docs/src/using-turing/compiler.md

```diff
@@ -0,0 +1,226 @@
+---
+title: Turing Compiler Design
+---
+
+In this section, I will describe the current design of Turing's model compiler which enables Turing to perform various types of Bayesian inference without changing the model definition. What we call "compiler" is essentially just a macro that transforms the user's code to something that Julia's dispatch can operate on and that Julia's compiler can successfully do type inference on for efficient machine code generation.
+
+# Overview
+
+The following terminology will be used in this section:
+
+- `D`: observed data variables conditioned upon in the posterior,
+- `P`: parameter variables distributed according to the prior distributions, these will also be referred to as random variables,
+- `Model`: a fully defined probabilistic model with input data, and
+- `Model` generator: a function that can be used to instantiate a `Model` instance by inputing data `D`.
+
+`Turing`'s `@model` macro defines a `Model` generator that can be used to instantiate a `Model` by passing in the observed data `D`.
+
+# `@model` macro
+
+The following are the main jobs of the `@model` macro:
+1. Parse `~` lines, e.g. `y ~ Normal(c*x, 1.0)`
+2. Figure out if a variable belongs to the data `D` and or to the parameters `P`
+3. Enable the handling of missing data variables in `D` when defining a `Model` and treating them as parameter variables in `P` instead
+4. Enable the tracking of random variables using the data structures `VarName` and `VarInfo`
+5. Change `~` lines with a variable in `P` on the LHS to a call to an `assume`-block
+6. Change `~` lines with a variable in `D` on the LHS to a call to an `observe`-block
+7. Enable type stable automatic differentiation of the model using type parameters
+
+Let's take the following model as an example:
+```julia
+@model gauss(x = missing, y = 1.0, ::Type{TV} = Vector{Float64}) where {TV <: AbstractVector} = begin
+    if ismissing(x)
+        x = TV(undef, 3)
+    end
+    p = TV(undef, 2)
+    p[1] ~ InverseGamma(2, 3)
+    p[2] ~ Normal(0, 1.0)
+    @. x[1:2] ~ Normal(p[2], sqrt(p[1]))
+    x[3] ~ Normal()
+    y ~ Normal(p[2], sqrt(p[1]))
+end
+```
+A `model::Model` can be defined using `gauss(rand(3), 1.0)` or `gauss(x = rand(3), y = 1.0)`. While constructing the model, if an argument is not passed in, it will be assigned to its default value. If there is no default value given, an error will be thrown. If an argument has a default value `missing`, when not passed in, it will be treated as a random variable. For variables which require an intialization because we need to loop or broadcast over its elements, such as `x` above, the following needs to be done:
+```julia
+if ismissing(x)
```

Comment:
Is there a particular reason for recommending `ismissing(x)` instead of `x === missing`? Even though it might seem they should be equivalent, they're actually not (see, e.g., https://github.com/JuliaLang/julia/issues/27681#issuecomment-399054971; I just reproduced the example there on Julia 1.3).

Reply:
I will recommend `===`, thanks!